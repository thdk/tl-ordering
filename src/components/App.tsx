import React, { useEffect, ReactNode } from 'react';
import OrderList from './OrderList';
import { IOrder } from '../interfaces/orders';
import { ThunkDispatch } from 'redux-thunk';
import { fetchOrders } from '../actions/orders';
import { connect } from 'react-redux';
import { IState } from '../interfaces/state';

type PropsFromState = {
  orders: IOrder[];
  isLoading: boolean;
};

type PropsFromDispatch = {
  fetchOrders: () => void;
}

export interface IAppProps {
  children: ReactNode;
}

type Props = PropsFromState & PropsFromDispatch & IAppProps;

const App = (props: Props) => {
  const { orders, fetchOrders, children, isLoading } = props;

  if (!orders.length)
    useEffect(() => {
      fetchOrders()
    }, []);

  const appJSX = isLoading ? <>LOADING</> : children;
  return (
    <div>
      {appJSX}
    </div>
  );
};

const mapStateToProps = (state: IState): PropsFromState => {
  return {
    orders: state.orders.orders,
    isLoading: state.orders.isLoading
  }
};

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>): PropsFromDispatch => {
  return {
    fetchOrders: () => dispatch(fetchOrders()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(App);