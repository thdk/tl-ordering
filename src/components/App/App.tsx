import React, { ReactNode, useEffect } from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";

import { getIsLoading } from "../../core/app/selectors";
import { IState } from "../../core/app/types";
import { fetchOrders as fetchOrderAction } from "../../core/orders/actions";
import { fetchProducts as fetchProductsAction } from "../../core/products/actions";

export interface IAppProps {
  children: ReactNode;
}

type Props = StateProps & DispatchProps & IAppProps;

export const App = (props: Props) => {
  const { children, isLoading, fetchOrders, fetchProducts } = props;

  useEffect(() => {
    fetchOrders();
    fetchProducts();
  }, []);

  const appJSX = isLoading ? <span data-testid="app-loading">LOADING</span> : children;
  return (
    <div>
      {appJSX}
    </div>
  );
};

type StateProps = ReturnType<typeof mapStatetoProps>;

const mapStatetoProps = (state: IState) => ({
  isLoading: getIsLoading(state),
});

type DispatchProps = ReturnType<typeof mapDispatchToProps>;

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      fetchProducts: fetchProductsAction,
      fetchOrders: fetchOrderAction,
    },
    dispatch,
  );

export default connect(mapStatetoProps, mapDispatchToProps)(App);
