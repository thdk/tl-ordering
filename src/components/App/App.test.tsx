import { fireEvent, getByTestId, render } from "@testing-library/react";
import React from "react";

import { shallow } from "enzyme";
import { App } from "./App";

const createProps = (isLoading: boolean) => ({
    fetchProducts: jest.fn(),
    fetchOrders: jest.fn(),
    isLoading,
});

describe("App rendering", () => {
    describe("when state is loading", () => {
        it("should match snapshot", () => {
            const component = shallow(
                <App {...createProps(false)}>App</App>,
            );

            expect(component.debug()).toMatchSnapshot();
        });

        it("should render loading indicator", () => {
            const { container } = render(<App {...createProps(true)}>App</App>);
            const loadingIndicator = getByTestId(container, "app-loading");
            expect(loadingIndicator).not.toBe(null);
        });
    });

    describe("when state is not loading", () => {
        it("should match snapshot", () => {
            const component = shallow(
                <App {...createProps(false)}>App</App>,
            );

            expect(component.debug()).toMatchSnapshot();
        });
    });
});
