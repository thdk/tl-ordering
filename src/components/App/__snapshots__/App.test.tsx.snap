// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering when state is loading should match snapshot 1`] = `
"<div>
  App
</div>"
`;

exports[`App rendering when state is not loading should match snapshot 1`] = `
"<div>
  App
</div>"
`;
